src/solvers/smt2/letify.cpp

```diff
@@ -46,26 +46,29 @@ void letifyt::collect_bindings(
   let_order.push_back(expr);
 }
 
-exprt letifyt::letify_rec(
+exprt letifyt::letify(
   const exprt &expr,
-  std::vector<exprt> &let_order,
-  const seen_expressionst &map,
-  std::size_t i)
+  const std::vector<exprt> &let_order,
+  const seen_expressionst &map)
 {
-  if(i >= let_order.size())
-    return substitute_let(expr, map);
+  exprt result = substitute_let(expr, map);
 
-  exprt current = let_order[i];
-  INVARIANT(
-    map.find(current) != map.end(), "expression should have been seen already");
+  // go backwards in let order
+  for(auto r_it = let_order.rbegin(); r_it != let_order.rend(); r_it++)
+  {
+    const exprt &current = *r_it;
 
-  if(map.find(current)->second.count < LET_COUNT)
-    return letify_rec(expr, let_order, map, i + 1);
+    auto m_it = map.find(current);
+    INVARIANT(m_it != map.end(), "expression should have been seen already");
+
+    if(m_it->second.count >= LET_COUNT)
```

Comment:
It would be nice to document this magic number. At least I have absolutely no idea why "2" is the right value, or indeed what it even is good for.

Reply:
Two is really the only viable choice, and the symbolic name could well be replaced by 2.
"2" means that the expression is used more than once.
"3" (or higher) implies that the procedure becomes worst-case exponential.

Reply:
Maybe using "1" instead of "2" would make things a lot clearer? That would make clear we are talking about something we have seen _more than once._ The named constant here really obfuscates this code, because 1) the name does not convey useful information and 2) one has to look elsewhere to find the value.

Reply:
Adding to my comment moments ago: I think this one should be `if(m_it->second.count != 1)`.

Reply:
... and the other one in `substitute_let` is likely also best written as `!= 1`, mirroring this one.